graph/network/diffusion_test.go

```diff
@@ -0,0 +1,495 @@
+			0, -1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
+			0, 0, 0, 1, -1. / 3., 0, 0, 0, 0, 0, 0,
+			0, 0, 0, 0, 1, -0.5, -0.5, -0.5, -0.5, -1, -1,
+			0, 0, 0, 0, -1. / 3., 1, 0, 0, 0, 0, 0,
+			0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
+			0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
+			0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
+			0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
+			0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
+		}),
+	},
+}
+
+func TestRandomWalkLaplacian(t *testing.T) {
+	const tol = 1e-14
+	for i, test := range randomWalkLaplacianTests {
+		g := simple.NewDirectedGraph()
+		for u, e := range test.g {
+			// Add nodes that are not defined by an edge.
+			if !g.Has(simple.Node(u)) {
+				g.AddNode(simple.Node(u))
+			}
+			for v := range e {
+				g.SetEdge(simple.Edge{F: simple.Node(u), T: simple.Node(v)})
+			}
+		}
+		l := NewRandomWalkLaplacian(g, test.damp)
+		_, c := l.Dims()
+		for j := 0; j < c; j++ {
+			if got := mat.Sum(l.Matrix.(*mat.Dense).ColView(j)); !floats.EqualWithinAbsOrRel(got, 0, tol, tol) {
+				t.Errorf("unexpected column sum for test %d, column %d: got:%v want:0", i, j, got)
+			}
+		}
+		l = NewRandomWalkLaplacian(sortedNodeGraph{g}, test.damp)
+		if !mat.EqualApprox(l, test.want, tol) {
+			t.Errorf("unexpected result for test %d:\ngot:\n% .2v\nwant:\n% .2v",
+				i, mat.Formatted(l), mat.Formatted(test.want))
+		}
+	}
+}
+
+type sortedNodeGraph struct {
+	graph.Graph
+}
+
+func (g sortedNodeGraph) Nodes() []graph.Node {
+	n := g.Graph.Nodes()
+	sort.Sort(ordered.ByID(n))
+	return n
+}
+
+var diffuseToEquilibriumTests = []struct {
+	g       []set
+	builder builder
+	h       map[int64]float64
+	damp    float64
+	tol     float64
+	iter    int
+
+	want   map[int64]float64
+	wantOK bool
+}{
+	{
+		g:       grid(5),
+		builder: simple.NewUndirectedGraph(),
+		h:       map[int64]float64{0: 1},
+		damp:    0.85,
+		tol:     1e-6,
+		iter:    1e4,
+
+		want: map[int64]float64{
+			A: 0.025000, B: 0.037500, C: 0.037500, D: 0.037500, E: 0.025000,
+			F: 0.037500, G: 0.050000, H: 0.050000, I: 0.050000, J: 0.037500,
+			K: 0.037500, L: 0.050000, M: 0.050000, N: 0.050000, O: 0.037500,
+			P: 0.037500, Q: 0.050000, R: 0.050000, S: 0.050000, T: 0.037500,
+			U: 0.025000, V: 0.037500, W: 0.037500, X: 0.037500, Y: 0.025000,
+		},
+		wantOK: true,
+	},
+	{
+		// Example graph from http://en.wikipedia.org/wiki/File:PageRanks-Example.svg 16:17, 8 July 2009
+		g: []set{
+			A: nil,
+			B: linksTo(C),
+			C: linksTo(B),
+			D: linksTo(A, B),
+			E: linksTo(D, B, F),
+			F: linksTo(B, E),
+			G: linksTo(B, E),
+			H: linksTo(B, E),
+			I: linksTo(B, E),
+			J: linksTo(E),
+			K: linksTo(E),
+		},
+		builder: simple.NewDirectedGraph(),
+		h: map[int64]float64{
+			A: 1. / 11.,
+			B: 1. / 11.,
+			C: 1. / 11.,
+			D: 1. / 11.,
+			E: 1. / 11.,
+			F: 1. / 11.,
+			G: 1. / 11.,
+			H: 1. / 11.,
+			I: 1. / 11.,
+			J: 1. / 11.,
+			K: 1. / 11.,
+		},
+		damp: 0.85,
+		tol:  1e-6,
+		iter: 1e4,
+
+		// This does not look like Page Rank because we do not
```

Comment:
Embedding homework problems in our source code? :-)

Reply:
Yes.